app/src/main/java/cm/aptoide/pt/app/AppViewManager.java

```diff
@@ -533,50 +540,103 @@ private void sendAdsBlockByOfferEvent() {
-                          viewModel.getId(), viewModel.getPackageName(), viewModel.getMd5sum(),
-                          viewModel.getVersionCode(), viewModel.getVersionName(),
-                          viewModel.getPath(), viewModel.getPathAlt(), viewModel.getObb(),
-                          viewModel.getAppcValue(), walletInstalled != null,
-                          viewModel.shouldShowOffer(), appViewAppInstalled != null,
-                          viewModel.getSize())))));
+      return promotionsManager.getPromotionsForPackage(cachedApp.getPackageName())
+          .filter(promotions -> !promotions.isEmpty())
+          .flatMap(promotionList -> {
+            Observable<List<Promotion>> promObs = Observable.just(promotionList);
+            Observable<WalletApp> walletApp = loadWallet();
+            Observable<DownloadModel> appDownloadModel = appViewAppDownloadModel();
+            Observable<DetailedApp> appViewModel = Observable.just(cachedApp);
+            Observable<Boolean> isAppMigrated =
+                appcMigrationService.isAppMigrated(cachedApp.getPackageName());
+            return Observable.combineLatest(promObs, walletApp, appDownloadModel, appViewModel,
+                isAppMigrated,
+                (proms, wallet, appDLM, appVM, migrate) -> mergeToPromotionViewModel(wallet, proms,
+                    appDLM, appVM, migrate));
+          })
+          .doOnNext(promotionViewModel -> cachedPromotionViewModel = promotionViewModel)
+          .switchIfEmpty(promoViewModelObs);
     }
   }
 
-  private WalletPromotionViewModel mapToWalletPromotion(List<PromotionApp> apps) {
-    if (apps == null || apps.isEmpty()) {
-      return new WalletPromotionViewModel(false);
-    } else {
-      PromotionApp app = apps.get(0);
-      if (app.getPackageName()
-          .equals("com.appcoins.wallet")) {
-        if (app.isClaimed()) {
-          promotionStatus = PromotionStatus.CLAIMED;
-          return new WalletPromotionViewModel(false);
-        } else {
-          promotionStatus = PromotionStatus.NOT_CLAIMED;
-          return new WalletPromotionViewModel(null, app.getName(), app.getAppIcon(), app.getAppId(),
-              app.getPackageName(), app.getMd5(), app.getVersionCode(), app.getVersionName(),
-              app.getDownloadPath(), app.getAlternativePath(), app.getObb(),
-              Math.round(app.getAppcValue()), false, true, false, app.getSize());
-        }
-      } else {
-        return new WalletPromotionViewModel(false);
-      }
-    }
+  /**
+   * If the user clicks before cachedPromotionViewModel is set, this will return false, even though
+   * it might be true later. We could call loadPromotionViewModel(), but in this case it will do a
+   * lot of work twice. I think it's acceptable to assume it's false if the promotion wasn't shown
+   * yet.
+   */
+  public boolean hasClaimablePromotion(Promotion.ClaimAction action) {
```

Comment:
We've talked about this, but just to sum it up: this might take a long while because there's two different requests for promotions (the get promotion and the get wallet app), so this might add a significant delay to the AppView load.

As it's stated in the method doc, I don't think it's a big deal. If the user clicks on install without first seeing the promotion, should we even count it? As it stands, the user wouldn't know that there's a promotion so it's weird to count it as such.